java/org/contikios/cooja/Simulation.java

```diff
@@ -150,68 +155,133 @@ public Simulation(Cooja cooja, long seed) {
     this.cooja = cooja;
     randomGenerator = new SafeRandom(this);
     randomSeed = seed;
-  }
-
-  @Override
-  public void run() {
-    assert isRunning : "Did not set isRunning before starting";
-    lastStartRealTime = System.currentTimeMillis();
-    lastStartSimulationTime = getSimulationTimeMillis();
-    speedLimitLastRealtime = lastStartRealTime;
-    speedLimitLastSimtime = lastStartSimulationTime;
+    this.simulationThread = new Thread(() -> {
+      boolean isSimulationRunning = false;
+      boolean isAlive = true;
+      do {
+        EventQueue.Pair nextEvent = null;
+        try {
+          while (isAlive) {
+            Runnable cmd;
+            do {
+              cmd = isSimulationRunning ? commandQueue.poll() : commandQueue.take();
+              if (cmd == null) {
+                // No command to execute
+              } else if (cmd == START) {
+                if (!isSimulationRunning) {
```

Comment:
`setRunning` looks safe so I think these if-statements can be skipped, if someone manages to pass two consecutive starts/stops they get the timestamp of the last/first one.

Reply:
I will update this. Thanks!